File: backend/app/schemas/starter_wardrobe.py
"""Phase 5d — starter wardrobe request/response shapes.

Templates expose only the metadata the client needs to render a "this kit
is for you" preview. Item-level details live inside the materialized
`wardrobe_items` rows the client fetches via /wardrobe.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class StarterWardrobeTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    template_id: str
    name: str
    gender: str | None
    age_range: str | None
    style_profile: str | None
    total_items: int
    items: list[dict[str, Any]]
    is_active: bool
    version: int


class TemplatesListResponse(BaseModel):
    templates: list[StarterWardrobeTemplateResponse]


class AssignStarterWardrobeRequest(BaseModel):
    """Optional explicit override. When omitted the server picks based on the
    user's shopping_style + age_range; missing/unsupported combinations fall
    through to a neutral default."""

    template_id: str | None = None


class UserStarterWardrobeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    template_id: UUID
    is_active: bool
    assigned_at: datetime
    deactivated_at: datetime | None
    deactivation_reason: str | None


class TransitionTrackingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    real_items_count: int
    starter_items_count: int
    percentage_real: float
    blending_ratio: float
    last_updated: datetime


class AssignStarterWardrobeResponse(BaseModel):
    """Returns the assignment, the chosen template's `template_id` slug, the
    count of items materialized this call (0 if no-op), and the live
    transition row so the client can render progress immediately."""

    assignment: UserStarterWardrobeResponse
    template_id: str
    items_materialized: int
    swapped: bool
    transition: TransitionTrackingResponse


class DeactivateStarterWardrobeRequest(BaseModel):
    reason: str | None = None


class DeactivateStarterWardrobeResponse(BaseModel):
    assignment: UserStarterWardrobeResponse

File: mobile/lib/modules/onboarding/models/starter_wardrobe.dart
/// DTOs for the starter-wardrobe endpoints.
///
/// The onboarding flow assigns a starter wardrobe so a brand-new user has
/// outfits before adding their own items; the backend auto-picks a template
/// from the user's shopping_style + age_range (neutral default otherwise) and
/// materializes its items into `/wardrobe`.
library;

/// One row of `GET /starter-wardrobe/templates` (the browsable catalogue).
class StarterTemplate {
  const StarterTemplate({
    required this.id,
    required this.templateId,
    required this.name,
    required this.totalItems,
    this.gender,
    this.ageRange,
    this.styleProfile,
  });

  final String id; // UUID
  final String templateId; // slug, e.g. neutral_default
  final String name;
  final int totalItems;
  final String? gender;
  final String? ageRange;
  final String? styleProfile;

  factory StarterTemplate.fromJson(Map<String, dynamic> json) {
    return StarterTemplate(
      id: json['id'] as String,
      templateId: json['template_id'] as String,
      name: json['name'] as String,
      totalItems: json['total_items'] as int,
      gender: json['gender'] as String?,
      ageRange: json['age_range'] as String?,
      styleProfile: json['style_profile'] as String?,
    );
  }
}

/// Result of `POST /starter-wardrobe/assign` — the chosen template slug, how
/// many items were materialized this call, and the live starter-item count.
class StarterWardrobeResult {
  const StarterWardrobeResult({
    required this.templateId,
    required this.itemsMaterialized,
    required this.swapped,
    required this.starterItemsCount,
  });

  final String templateId;
  final int itemsMaterialized;
  final bool swapped;
  final int starterItemsCount;

  factory StarterWardrobeResult.fromJson(Map<String, dynamic> json) {
    final transition = json['transition'] as Map<String, dynamic>?;
    return StarterWardrobeResult(
      templateId: json['template_id'] as String,
      itemsMaterialized: json['items_materialized'] as int? ?? 0,
      swapped: json['swapped'] as bool? ?? false,
      starterItemsCount: transition?['starter_items_count'] as int? ?? 0,
    );
  }

  /// Best count to show the user: items added this call, else the live total.
  int get displayCount =>
      itemsMaterialized > 0 ? itemsMaterialized : starterItemsCount;
}
